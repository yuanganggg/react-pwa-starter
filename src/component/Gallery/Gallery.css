/* standard slider selector */
.slider-box {
  position: relative;
  width: 100%;
  height: calc(100vh - 46px);
  overflow: hidden;
}

.slider-box-item {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  will-change: transform;
}

.slider-box-image {
  width: 100%;
  height: 100%;
  object-fit: cover;
}

/* slide animation, need global selector */
:global .slide-enter {
  opacity: 0;
  transform: translateX(100%);
}

:global .slide-enter.slide-enter-active {
  opacity: 1;
  transform: translateX(0);
  transition: all 400ms ease-in-out;
}

:global .slide-exit {
  opacity: 1;
  transform: translateX(0);
}

:global .slide-exit.slide-exit-active {
  opacity: 0;
  transform: translateX(-100%);
  transition: all 500ms ease-in-out;
}
