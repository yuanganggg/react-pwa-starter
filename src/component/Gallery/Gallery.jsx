/**
 * @description - Gallery slider
 * @author - huang.jian <hjj491229492@hotmail.com>
 */
// External
import React, { Component } from 'react';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import { Observable } from 'rxjs/Observable';
import 'rxjs/add/observable/timer';
import 'rxjs/add/operator/map';

// Style
import './Gallery.scss';

// Internal
import cartoon1 from '../../assets/1.jpg';
import cartoon2 from '../../assets/2.jpg';
import cartoon3 from '../../assets/3.jpg';
import cartoon4 from '../../assets/4.jpg';
import cartoon5 from '../../assets/5.jpg';

// Component Scope
const cartoons = [cartoon1, cartoon2, cartoon3, cartoon4, cartoon5];

export default class Gallery extends Component {
  constructor(props) {
    super(props);

    this.state = {
      activeImage: [cartoon1]
    };

    // Transition timeout
    this.timeout = {
      enter: 400,
      exit: 500
    };
  }

  componentWillMount() {
    this.state$ = Observable.timer(500, 2000)
      .map((count) => ({
        activeImage: [cartoons[count % 5]]
      }));
  }

  componentDidMount() {
    this.subscription = this.state$.subscribe((state) => {
      this.setState(state);
    });
  }

  componentWillUnmount() {
    this.subscription.unsubscribe();
  }

  render() {
    return (
      <article>
        <div className="slider-box">
          <TransitionGroup>
            {
              this.state.activeImage.map((url) => (
                <CSSTransition
                  key={url}
                  timeout={this.timeout}
                  classNames="slide"
                >
                  <div className="slider-box__item">
                    <img className="slider-box__image" src={url} alt="cartoon" />
                  </div>
                </CSSTransition>
              ))
            }
          </TransitionGroup>
        </div>
      </article>
    );
  }
}
