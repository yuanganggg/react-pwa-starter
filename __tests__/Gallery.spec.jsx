/**
 * @description - jest configuration
 * @author - bornkiller <hjj491229492@hotmail.com>
 */
// External
import React from 'react';
import { shallow } from 'enzyme';

// Internal
import Gallery from '../src/component/Gallery';

describe('Gallery component suits', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it('should render image slider', () => {
    const wrapper = shallow(<Gallery />);

    // manual trigger lifecycle
    wrapper.instance().componentDidMount();
    expect(wrapper.text()).toMatchSnapshot();

    // manual trigger lifecycle
    wrapper.instance().componentWillUnmount();
    expect(wrapper.text()).toMatchSnapshot();
  });
});

