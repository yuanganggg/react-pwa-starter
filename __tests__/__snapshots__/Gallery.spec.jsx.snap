// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery component suits should render image slider 1`] = `"<TransitionGroup />"`;

exports[`Gallery component suits should render image slider 2`] = `"<TransitionGroup />"`;
